### PR TITLE
Filter Subscriptions #97

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

File: src/main/resources/lib/guillotine/subscription/event.js [hunk 1]
```diff
@@ -27,17 +28,52 @@ function createEventObjectType(context) {
...
+                        return true;
+                    }
+                }
+            });
         }
     });
     return processor;
 }
 
+function anyFilterMatch(event, siteRepo, type) {
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/subscription/event.js:66]: This is a confusing name. Which of  _any_ filters this function matches?

File: src/main/resources/lib/guillotine/index.js [hunk 2]
```diff
@@ -65,10 +67,83 @@ function createContext(options) {
...
+        }
+    } catch (e) {
+        log.error('Error while handling Start GraphQL-WS message', e);
+        throw e;
+    }
+}
+
+function initWebSockets(schema) {
+    return function (event) {
+        switch (event.type) {
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:118]: need default

File: src/main/resources/lib/guillotine/index.js [hunk 3]
```diff
@@ -65,10 +67,83 @@ function createContext(options) {
...
+        case 'close':
+            cancelSubscription(event.session.id);
+            break;
+        case 'message':
+            var message = JSON.parse(event.message);
+            switch (message.type) {
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:124]: need default

Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:124]: switch/case in switch/case is hard to read

File: src/main/resources/lib/guillotine/index.js [hunk 4]
```diff
@@ -65,10 +67,83 @@ function createContext(options) {
...
+            const subscriber = graphQlRxLib.createSubscriber({
+                onNext: (result) => {
+                    webSocketLib.send(sessionId, JSON.stringify({
+                        type: 'data',
+                        id: graphQLOperationId,
+                        payload: result
+                    }));
+                }
+            });
+            Java.synchronized(() => graphQlSubscribers[sessionId] = subscriber, graphQlSubscribers)();
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:107]: Consider using socket groups instead

File: src/main/resources/lib/guillotine/index.js [hunk 5]
```diff
@@ -4,6 +4,8 @@ const dynamicLib = require('/lib/guillotine/dynamic/index');
 const graphQlLib = require('/lib/guillotine/graphql');
 const rootQueryLib = require('/lib/guillotine/query/root-query');
 const rootSubscriptionLib = require('/lib/guillotine/subscription/root-subscription');
+const graphQlRxLib = require('/lib/graphql-rx');
+const webSocketLib = require('/lib/xp/websocket');
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:8]: Where this dependency is declared

**Review by rymsha — CHANGES_REQUESTED** [review]
read single comments

File: src/main/resources/lib/guillotine/subscription/event.js (hunk 1 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/subscription/event.js]: done

File: src/main/resources/lib/guillotine/index.js (hunk 2 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/index.js:118]: done

File: src/main/resources/lib/guillotine/index.js (hunk 3 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/index.js]: done

Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/index.js]: done

File: src/main/resources/lib/guillotine/index.js (hunk 4 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/index.js:107]: left without changes after discuss

File: src/main/resources/lib/guillotine/index.js (hunk 5 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/index.js:8]: done

File: src/main/resources/lib/guillotine/index.js [hunk 6]
```diff
@@ -65,10 +67,93 @@ function createContext(options) {
...
     return context;
 }
 
+//──────────────────────────────────────────────────────────────────────────────
+// GraphQL WS Protocol
+//──────────────────────────────────────────────────────────────────────────────
+const graphQlSubscribers = {};
+
+function cancelSubscription(sessionId) {
+    Java.synchronized(() => {
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:81]: Create a separate issue, probably in XP. It should be possible to use websoket context for such needs without involving java

File: src/main/resources/lib/guillotine/index.js [hunk 7]
```diff
@@ -65,10 +67,93 @@ function createContext(options) {
     }
     context.options.applications = context.options.applications || [app.name];
     context.options.allowPaths = context.options.allowPaths || [];
+    context.options.subscriptionEventTypes = context.options.subscriptionEventTypes || ['node.*'];
 
     return context;
 }
 
+//──────────────────────────────────────────────────────────────────────────────
+// GraphQL WS Protocol
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:76]: Move to own file - in the lib-grphql?

**Review by rymsha — CHANGES_REQUESTED** [review]

File: src/main/resources/lib/guillotine/index.js (hunk 7 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/guillotine/index.js:76]: @rymsha I guess will be better to figure out the better solution for that issue and then move it to lib-graphQL if needed, but not now.

**Review by rymsha — APPROVED** [review]

**Pull request merged**